File: Thay Dong/Buoi 1/BUDGET.pas
Const
        maxn=10001;
        maxm=1000001;
        oo=trunc(1e17);
Var
        n,m,p,t:int64;
        visited:array[0..maxn] of int64;
        s,r:array[0..maxn] of int64;
        mark:array[0..maxm] of int64;
        toph:array[0..maxm] of longint;
        h:array[0..maxn,0..maxn] of int64;
Procedure nhap;
Begin
        Assign(input,'BUDGET.INP');
        reset(input);
        Assign(output,'BUDGET.OUT');
        rewrite(output);
End;
Function min(x,y:int64):int64;
Begin
        If x<y then exit(x) else exit(y);
End;
Function visit(u:int64):boolean;
Var
        i:longint;
Begin
        If visited[u]<>t then visited[u]:=t 
        else exit(false);
        For i:=1 to h[u,toph[u]] do
        If (mark[h[u,i]]=0)  or visit(mark[h[u,i]]) then
        Begin
                mark[h[u,i]]:=u;
                exit(true);
        End;
        exit(false);
End;
Function check(d:int64):boolean;
Var
        i,j,k:longint;
        cur:longint;
        kk:int64;
Begin
        For i:=1 to n do toph[i]:=0;
        Fillchar(mark,sizeof(mark),0);
        For i:=1 to n do
        For j:=1 to m do
        For k:=1 to n do
        Begin
                kk:=p mod s[j];
                If kk<>0 then kk:=1;
                kk:=kk+(p div s[j]);
                IF (r[i]+kk*k)<=d+1 then
                Begin
                        inc(toph[i]);
                        h[i,toph[i]]:=n*(j-1)+k;
                End;
        End;
        cur:=0;
        For i:=1 to n do
        Begin
                inc(t);
                If visit(i) then
                inc(cur);
        End;
        if (cur>=n) then exit(true);
        check:=false;
End;
Procedure tinh;
Var
        i:longint;
        d,c,g:int64;
        res:int64;
Begin
        readln(m,n,p);
        For i:=1 to m do read(s[i]);
        For i:=1 to n do read(r[i]);
        d:=0;
        c:=oo;
        res:=oo+1;
        t:=0;
        While d<=c do
        Begin
                g:=(d+c) div 2;
                IF check(g) then
                Begin
                        res:=min(res,g);
                        c:=g-1;
                End else d:=g+1;
        End;
        write(res);
End;
Begin
        nhap;
        tinh;
End.

File: Thay Dong/Buoi 1/BUDGET.cpp
#include <bits/stdc++.h>
#define FastIO ios::sync_with_stdio(0); cin.tie(NULL); cout.tie(NULL);
using namespace std;
const int MaxN=2003;
int64_t n, m, p, t=0;
int64_t visited[MaxN];
vector <int64_t> s, r, mark;
vector <vector<int64_t> > h;
bool visit(int64_t u) {
    if (visited[u]!=t) visited[u]=t;
    else return false;
    for (int i=0;i<h[u].size();i++) {
        if (!mark[h[u][i]] || visit(mark[h[u][i]])) {
            mark[h[u][i]]=u;
            return true;
        }
    }
    return false;
}
bool check(int64_t d) {
    for (int i=1;i<=n;i++) h[i].clear();
    for (int i=1;i<mark.size();i++) mark[i]=0;
    for (int i=1;i<=n;i++) for (int j=1;j<=m;j++)
        for (int64_t k=1;k<=n;k++)
        if ((r[i]+(p/s[j]+(p%s[j]!=0))*k)<=d+1) h[i].push_back(n*(j-1)+k);
    int cur=0;
    for (int i=1;i<=n;i++) {
        t++;
        cur+=visit(i);
    }
    return cur>=n;
}
int32_t main(void) {
    FastIO;
    freopen("BUDGET.INP","r",stdin);
    freopen("BUDGET.OUT","w",stdout);
    cin>>m>>n>>p; h.resize(n+1);
    s.resize(m+1); r.resize(n+1);
    mark.resize(m*n+n+1);
    for (int i=1;i<=m;i++) cin>>s[i];
    for (int i=1;i<=n;i++) cin>>r[i];
    int64_t l=0, r=1e18, res=1e18+1;
    while (l<=r) {
        int64_t m=(l+r)/2;
        if (check(m)) {
            res=min(m,res);
            r=m-1;
        } else l=m+1;
    }
    cout<<res;
    return 0;
}